source/Mlos.Python/mlos/Spaces/unit_tests/TestFilteringOutInvalidRows.py

```diff
@@ -0,0 +1,92 @@
+from mlos.Spaces import ContinuousDimension, DiscreteDimension, Point
+from mlos.Tracer import Tracer, trace, traced
+
+class TestFilteringOutInvalidRows(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        """ Sets up all the singletons needed to test the BayesianOptimizer.
+
+        """
+        warnings.simplefilter("error")
+        global_values.declare_singletons()
+        global_values.tracer = Tracer(actor_id=cls.__name__, thread_id=0)
+        cls.logger = create_logger(logger_name=cls.__name__)
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        temp_dir = os.path.join(os.getcwd(), "temp")
+        if not os.path.exists(temp_dir):
+            os.mkdir(temp_dir)
+        trace_output_path = os.path.join(temp_dir, f"{cls.__name__}.json")
+        print(f"Dumping trace to {trace_output_path}")
+        global_values.tracer.dump_trace_to_file(output_file_path=trace_output_path)
+
+    def test_filtering_out_invalid_rows(self):
+        spaces = [
+            BayesianOptimizerConfig.CONFIG_SPACE,
+            GlowWormSwarmOptimizerConfig.CONFIG_SPACE
+        ]
+
+        # Just to make sure we are testing both hierarchical and flat code paths.
+        #
+        self.assertTrue(any(space.is_hierarchical() for space in spaces))
+        self.assertTrue(any(not space.is_hierarchical() for space in spaces))
+
+        num_samples = 1000
+        for space in spaces:
+            random_dataframe_with_invalid_rows = space.random_dataframe(num_samples=num_samples)
+            for dimension in space.dimensions:
+                if isinstance(dimension, (ContinuousDimension, DiscreteDimension)):
+                    # This makes about half of the rows invalid.
+                    #
+                    random_dataframe_with_invalid_rows.loc[:, [dimension.name]] *= 2
+                    break
+
+            with traced(scope_name="slow_filtering"):
+                # Let's filter out invalid rows the slow way.
+                #
+                valid_indices = []
+                for idx in random_dataframe_with_invalid_rows.index:
+                    row_as_df = random_dataframe_with_invalid_rows.loc[[idx]]
+                    row_as_point = Point.from_dataframe(row_as_df)
+                    if row_as_point in space:
+                        valid_indices.append(idx)
+                slow_valid_rows_index = pd.Index(valid_indices)
+
+            print(f"{len(slow_valid_rows_index)}/{len(random_dataframe_with_invalid_rows.index)} rows are valid.")
+            self.assertTrue(0 < len(slow_valid_rows_index))
+            self.assertTrue(len(slow_valid_rows_index) < num_samples)
+
+            # Let's filter out invalid rows the fast way.
+            #
+            fast_valid_rows_index = space.filter_out_invalid_rows(original_dataframe=random_dataframe_with_invalid_rows, exclude_extra_columns=True).index

```

Comment:
Do you want to also validate that you expect this to be faster than the slow method as a sort of performance regression test?  Might be prone to flakiness or just be too long?

Reply:
No, this is a functional test.